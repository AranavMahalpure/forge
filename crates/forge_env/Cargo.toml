[package]
name = "forge_env"
version = "0.1.0"
edition = "2021"

[dependencies]
derive_more = { version = "1.0.0", features = ["from", "display"] }
handlebars = "6.2.0"
serde = "1.0.216"
dirs = "5.0.1"
