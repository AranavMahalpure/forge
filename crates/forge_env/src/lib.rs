mod env;
mod error;
pub use env::{Environment, EnvironmentValue};
pub use error::*;
