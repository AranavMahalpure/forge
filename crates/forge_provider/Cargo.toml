[package]
name = "forge_provider"
version = "0.1.0"
edition = "2021"

[dependencies]
tokio = { version = "1", features = ["full"] }
tokio-stream = "0.1"
reqwest = { version = "0.12", features = ["json", "rustls-tls"] }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
tracing = "0.1.41"
async-trait = "0.1.83"
derive_more = { version = "1.0.0", features = ["from", "display"] }
derive_setters = "0.1.6"
forge_tool = { path = "../forge_tool" }
reqwest-eventsource = "0.6.0"
moka2 = "0.13.0"
strum = "0.26.3"
strum_macros = "0.26.4"
nom = "8.0.0-beta.1"
utoipa = { version = "4.2.0", features = ["axum_extras"] }
