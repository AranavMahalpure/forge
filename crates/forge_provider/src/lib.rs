mod error;
mod ollama;
mod open_router;
mod provider;

pub use error::*;
pub use provider::*;

pub struct Service;
