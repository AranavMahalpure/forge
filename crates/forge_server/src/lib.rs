mod context;
mod error;
mod log;
mod prompts;
mod routes;
mod schema;
mod service;
mod template;

pub use error::*;
pub use routes::API;
pub use service::{ChatResponse, RootAPIService, Service};
