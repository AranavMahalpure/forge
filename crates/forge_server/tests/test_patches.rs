#[cfg(test)]
mod tests {
    use forge_env::Environment;
    use forge_provider::ModelId;
    use forge_server::{ChatRequest, ChatResponse, Server};
    use regex::Regex;
    use tokio_stream::StreamExt;

    macro_rules! assert {
        ($file_path:expr) => {
            for file_path in $file_path.iter() {
                let snap_name = snap_name(file_path);
                dbg!(&snap_name);
                let file_path = format!("{}/{}", env!("CARGO_MANIFEST_DIR"), file_path);
                let a = read(format!("{}", file_path));
                let re = Regex::new(r"\.(\w+)$").unwrap();
                let b = read(re.replace(&file_path, "_updated.$1"));

                let patch = diffy::create_patch(&a, &b);

                let mut patch_output = vec![];
                diffy::PatchFormatter::new()
                    .write_patch_into(&patch, &mut patch_output)
                    .expect("Failed to format patch");

                insta::assert_snapshot!(snap_name, String::from_utf8(patch_output).unwrap());
            }
            delete_updated_files(&$file_path);
        };
    }

    fn snap_name<T: AsRef<str>>(path: T) -> String {
        let path = path.as_ref();

        path.replace("/", "_").replace(".", "_")
    }

    async fn server() -> Server {
        let env = Environment::from_env().await.unwrap();
        let api_key = std::env::var("FORGE_KEY").expect("FORGE_KEY must be set");
        Server::new(env, api_key)
    }

    fn read<T: AsRef<str>>(path: T) -> String {
        std::fs::read_to_string(path.as_ref())
            .unwrap_or_else(|_| panic!("Failed to read file {}", path.as_ref()))
    }

    fn delete_updated_files<T: AsRef<str>>(file_path: &[T]) {
        for file in file_path {
            let file = format!(
                "{}/tests/{}",
                env!("CARGO_MANIFEST_DIR"),
                file.as_ref().replace(".md", "_updated.md")
            );
            let _ = std::fs::remove_file(file);
        }
    }

    async fn chat<T: AsRef<str>>(base: T, file_paths: &[T]) -> Vec<ChatResponse> {
        delete_updated_files(file_paths);
        let server = server().await;
        let req = ChatRequest::default()
            .message(format!("{} in the file(s) located at {} .Do not change the input file directly, create another file with changes within the same dir as the original file and name it <file_name>_updated.ext", base.as_ref(), file_paths.iter().map(|f| f.as_ref()).collect::<Vec<&str>>().join(", ")))
            .model(ModelId::new("anthropic/claude-3.5-haiku"));

        server
            .chat(req)
            .await
            .unwrap()
            .collect::<Vec<ChatResponse>>()
            .await
    }

    #[tokio::test]
    async fn test_rs_patches() {
        let file_paths = vec!["tests/fixtures/non_idiomatic.rs"];
        let _resp = chat("Refactor the following Rust code to use constructs like iterators and functional programming patterns. Avoid explicit loops and make the code concise while preserving its functionality. Do not rename variables.", &file_paths).await;
        assert!(file_paths);
    }
}
