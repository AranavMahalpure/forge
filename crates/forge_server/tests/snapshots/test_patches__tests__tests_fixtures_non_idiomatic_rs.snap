---
source: crates/forge_server/tests/test_patches.rs
expression: "String :: from_utf8(patch_output).unwrap()"
---
--- original
+++ modified
@@ -1,9 +1,7 @@
 fn main() {
-    let mut sum = 0;
-    for i in 1..=10 {
-        if i % 2 == 0 {
-            sum += i;
-        }
-    }
+    let sum: i32 = (1..=10)
+        .filter(|&i| i % 2 == 0)
+        .sum();
+    
     println!("Sum of even numbers: {}", sum);
-}
+}
\ No newline at end of file
